implement Literal::byte_character

Emilgardis opened the pull request:
without this, the only way to create a `LitKind::Byte` is by
doing `"b'a'".parse::<Literal>()`, this solves that by enabling
`Literal::byte_character(b'a')`

cc #71358

The tracking issue is #115268

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5292376305/jobs/9579188710) [(plain)](https://github.com/rust-lang/rust/commit/4c3fffc3b04cc6849b71b8283ba1611bc8bdc8a7/checks/14328649482/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully tagged rust-ci:latest
Built container sha256:09fd243785997cd73520582d86c93a2193cb0273b36283cf790ebda4cab056b6
Uploading finished image to https://ci-caches.rust-lang.org/docker/b0b14de1f8ccdae9d6691fe51748b78fc0caa47e31fb6a1c51fdc5b51c020f4f6b285907dbf35a56f7233a5f40c7dfe408e7d8869346843744b7d1b2da0c8c4e

<botocore.awsrequest.AWSRequest object at 0x7f8c032483d0>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
[CI_JOB_NAME=x86_64-gnu-llvm-14]
[CI_JOB_NAME=x86_64-gnu-llvm-14]
---
---- [ui] tests/ui/proc-macro/test.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/proc-macro/test.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/checkout/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/test" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/test/auxiliary"
stdout: none
--- stderr -------------------------------
error: non-ASCII character in byte literal
  --> <proc-macro source code>:1:3
LL | b'❤'
   |   ^
   |   |
   |   must be ASCII
   |   must be ASCII
   |   this multibyte character does not fit into a single byte

error: proc macro panicked
  --> fake-test-src-base/proc-macro/test.rs:10:1
   |
LL | proc_macro_api_tests::run!();
   |
   |
   = help: message: assertion failed: \"b\'❤\'\".parse::<Literal>().is_err()
error: aborting due to 2 previous errors
------------------------------------------


```

</details>

jackh726 commented:
r? libs-api

dtolnay commented on compiler/rustc_feature/src/active.rs:
```diff
@@ -489,6 +489,8 @@ declare_features! (
     (active, plugin, "1.0.0", Some(29597), None),
     /// Allows exhaustive integer pattern matching on `usize` and `isize`.
     (active, precise_pointer_size_matching, "1.32.0", Some(56354), None),
+    /// Allows creating a `b'b'` `Literal`
+    (active, proc_macro_byte_character, "CURRENT_RUSTC_VERSION", Some(71358), None),
```
I believe this list is only for compiler and language features. This entry shouldn't need to be here. The library stability attribute is sufficient.

dtolnay commented on library/proc_macro/src/lib.rs:
```diff
@@ -1357,6 +1357,13 @@ impl Literal {
     }
 
+    /// Byte character literal.
+    #[unstable(feature = "proc_macro_byte_character", issue = "71358")]
```
Please open a dedicated tracking issue.

Emilgardis commented:
@rustbot label +S-waiting-on-author -S-waiting-on-review

JohnCSimon commented:
@Emilgardis 
ping from triage - can you post your status on this PR and address the change requests? There hasn't been an update in a few months. Thanks!


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

Emilgardis commented:
I'll work on it today, thanks for the ping

Emilgardis commented:
@rustbot ready

dtolnay reviewed (approved):
Thank you!

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 5f2cec2608c1d0bc6d012283566eba70d8f211ce has been approved by `dtolnay`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=dtolnay 5f2cec2608c1d0bc6d012283566eba70d8f211ce -->
<!-- homu: {"type":"Approved","sha":"5f2cec2608c1d0bc6d012283566eba70d8f211ce","approver":"dtolnay","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 5f2cec2608c1d0bc6d012283566eba70d8f211ce with merge 45bc58cffbe18803ec17ba907763cc38bbd5a908...
<!-- homu: {"type":"BuildStarted","head_sha":"5f2cec2608c1d0bc6d012283566eba70d8f211ce","merge_sha":"45bc58cffbe18803ec17ba907763cc38bbd5a908"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6115933123/job/16600415309)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/6115933123/job/16600415309","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/6115933123/job/16600339746) [(plain)](https://github.com/rust-lang-ci/rust/commit/45bc58cffbe18803ec17ba907763cc38bbd5a908/checks/16600339746/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] std_detect test:false 0.659
[RUSTC-TIMING] hashbrown test:false 0.868
[RUSTC-TIMING] miniz_oxide test:false 1.048
[RUSTC-TIMING] core test:false 22.913
##[error]The runner has received a shutdown signal. This can happen when the runner service is stopped, or a manually started runner is canceled.
  local time: Fri Sep  8 00:01:55 UTC 2023

  network time: Fri, 08 Sep 2023 00:01:55 GMT
##[endgroup]
```

</details>

dtolnay commented:
I can't tell what failed. @bors retry

dtolnay commented:
@bors retry

bors commented:
:hourglass: Testing commit 5f2cec2608c1d0bc6d012283566eba70d8f211ce with merge 679b53f5937ef4d3dc4ead9f3591878d3e33fdf0...
<!-- homu: {"type":"BuildStarted","head_sha":"5f2cec2608c1d0bc6d012283566eba70d8f211ce","merge_sha":"679b53f5937ef4d3dc4ead9f3591878d3e33fdf0"} -->

rust-log-analyzer commented:
The job **`test-various`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/6116712539/job/16602319120) [(plain)](https://github.com/rust-lang-ci/rust/commit/679b53f5937ef4d3dc4ead9f3591878d3e33fdf0/checks/16602319120/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
2   --> $DIR/feature-gate-proc_macro_byte_character.rs:8:5
3    |

7    = note: see issue #115268 <https://github.com/rust-lang/rust/issues/115268> for more information
8    = help: add `#![feature(proc_macro_byte_character)]` to the crate attributes to enable
- error: aborting due to previous error
+ error: aborting due to previous error; 1 warning emitted
11 
12 For more information about this error, try `rustc --explain E0658`.
---

error[E0658]: use of unstable library feature 'proc_macro_byte_character'
##[error]  --> /checkout/tests/ui/feature-gates/feature-gate-proc_macro_byte_character.rs:8:5
   |
LL |     Literal::byte_character(b'a'); //~ ERROR use of unstable library feature 'proc_macro_byte_character'
   |
   = note: see issue #115268 <https://github.com/rust-lang/rust/issues/115268> for more information
   = note: see issue #115268 <https://github.com/rust-lang/rust/issues/115268> for more information
   = help: add `#![feature(proc_macro_byte_character)]` to the crate attributes to enable
error: aborting due to previous error; 1 warning emitted

For more information about this error, try `rustc --explain E0658`.
------------------------------------------
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6116712539/job/16602645420)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/6116712539/job/16602645420","builder_name":"checks-actions"} -->

dtolnay commented:
@rustbot author

dtolnay reviewed (changes requested):
I think it would be fine to delete that ui test.

dtolnay commented:
Or, restrict it to some non-wasm architecture only.

https://github.com/rust-lang/rust/blob/5adddad28c6c6ae467061ca22dd9819bb0069d1a/tests/ui/feature-gates/feature-gate-asm_const.rs#L1

Emilgardis commented:
> Or, restrict it to some non-wasm architecture only.
> 
> https://github.com/rust-lang/rust/blob/5adddad28c6c6ae467061ca22dd9819bb0069d1a/tests/ui/feature-gates/feature-gate-asm_const.rs#L1

sounds good, was looking for a `only-host-tools` or similar

ehuss commented:
@bors r-

synchronzing the queue

Emilgardis committed:
implement Literal::byte_character

without this, the only way to create a `LitKind::Byte` is by
doing `"b'a'".parse::<Literal>()`, this solves that by enabling
`Literal::byte_character(b'a')`

Emilgardis commented:
@rustbot ready

i think `// force-host` solves my issue, can't reproduce with `--target wasm32-unknown-unknown`

dtolnay reviewed (approved):
Thanks!

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 74f52613454714cd033165cc635a407ef61f04c2 has been approved by `dtolnay`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=dtolnay 74f52613454714cd033165cc635a407ef61f04c2 -->
<!-- homu: {"type":"Approved","sha":"74f52613454714cd033165cc635a407ef61f04c2","approver":"dtolnay","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 74f52613454714cd033165cc635a407ef61f04c2 with merge acfb46db5a01fec939a28f2cd80f24289d903d71...
<!-- homu: {"type":"BuildStarted","head_sha":"74f52613454714cd033165cc635a407ef61f04c2","merge_sha":"acfb46db5a01fec939a28f2cd80f24289d903d71"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6286396300/job/17070634868)
Approved by: dtolnay
Pushing acfb46db5a01fec939a28f2cd80f24289d903d71 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"dtolnay","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6286396300/job/17070634868"},"merge_sha":"acfb46db5a01fec939a28f2cd80f24289d903d71"} -->

rust-timer commented:
Finished benchmarking commit (acfb46db5a01fec939a28f2cd80f24289d903d71): [comparison URL](https://perf.rust-lang.org/compare.html?start=13e6f24b9adda67852fb86538541adaa68aff6e8&end=acfb46db5a01fec939a28f2cd80f24289d903d71&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=13e6f24b9adda67852fb86538541adaa68aff6e8&end=acfb46db5a01fec939a28f2cd80f24289d903d71&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=13e6f24b9adda67852fb86538541adaa68aff6e8&end=acfb46db5a01fec939a28f2cd80f24289d903d71&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.9%  | [1.4%, 2.2%]   | 4     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.8% | [-2.8%, -2.8%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=13e6f24b9adda67852fb86538541adaa68aff6e8&end=acfb46db5a01fec939a28f2cd80f24289d903d71&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.6% | [2.6%, 2.6%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=13e6f24b9adda67852fb86538541adaa68aff6e8&end=acfb46db5a01fec939a28f2cd80f24289d903d71&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 631.479s -> 629.795s (-0.27%)
**Artifact size**: 317.13 MiB -> 317.18 MiB (0.01%)
